File: python/chapter04/prog.py
def find_max_crossing_subarray(a, low, mid, high):
    left_sum = -float('inf')
    sumv = 0
    for i in range(mid, low - 1, -1):
        sumv += a[i]
        if sumv > left_sum:
            left_sum = sumv
            max_left = i

    right_sum = -float('inf')
    sumv = 0
    for j in range(mid + 1, high + 1):
        sumv += a[j]
        if sumv > right_sum:
            right_sum = sumv
            max_right = j

    return max_left, max_right, left_sum + right_sum


def find_maximum_subarray(a, low, high):
    if high == low:
        return low, high, a[low]
    else:
        mid = (low+high) // 2
        left_low, left_high, left_sum = find_maximum_subarray(
            a, low, mid)
        right_low, right_high, right_sum = find_maximum_subarray(
            a, mid + 1, high)
        cross_low, cross_high, cross_sum = find_max_crossing_subarray(
            a, low, mid, high)

        if left_sum >= right_sum and left_sum >= cross_sum:
            return left_low, left_high, left_sum
        elif right_sum >= left_sum and right_sum >= cross_sum:
            return right_low, right_high, right_sum
        else:
            return cross_low, cross_high, cross_sum


def square_matrix_multiply(a, b):
    n = len(a)
    c = [[0] * n for i in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                c[i][j] += a[i][k] * b[k][j]
    return c

File: go/chapter04/prog.go
package chapter04

import (
	"math"
)

func findMaxCrossingSubarray(a []float64, low, mid, high int) (int, int, float64) {
	leftSum := math.Inf(-1)
	sum := float64(0)
	maxLeft := mid
	for i := mid; i >= low; i-- {
		sum += a[i]
		if sum > leftSum {
			leftSum = sum
			maxLeft = i
		}
	}

	rightSum := math.Inf(-1)
	sum = float64(0)
	maxRight := mid + 1
	for j := mid + 1; j <= high; j++ {
		sum += a[j]
		if sum > rightSum {
			rightSum = sum
			maxRight = j
		}
	}

	return maxLeft, maxRight, leftSum + rightSum
}

func findMaximumSubarray(a []float64, low, high int) (int, int, float64) {
	if high == low {
		return low, high, a[low]
	}
	mid := (low + high) / 2
	leftLow, leftHigh, leftSum := findMaximumSubarray(a, low, mid)
	rightLow, rightHigh, rightSum := findMaximumSubarray(a, mid+1, high)
	crossLow, crossHigh, crossSum := findMaxCrossingSubarray(a, low, mid, high)

	if leftSum >= rightSum && leftSum >= crossSum {
		return leftLow, leftHigh, leftSum
	} else if rightSum >= leftSum && rightSum >= crossSum {
		return rightLow, rightHigh, rightSum
	} else {
		return crossLow, crossHigh, crossSum
	}
}

func squareMatrixMultiply(a, b [][]float64) [][]float64 {
	n := len(a)
	c := make([][]float64, n)
	for i := 0; i < n; i++ {
		c[i] = make([]float64, n)
		for j := 0; j < n; j++ {
			for k := 0; k < n; k++ {
				c[i][j] += a[i][k] * b[k][j]
			}
		}
	}
	return c
}
